### PR TITLE
Implement a link state monitor task

### LINKED ISSUE #37: Make a link state monitor task
The link state monitor basically:

1. Collects the complete list of all ports on a device, including their link state
2. Detects link state changes since the last collection and makes new events for changed states

### Implementation breakdown

- [x] #69
- [x] #70
- [x] #75
- [x] #77
- [x] #79

### LINKED ISSUE #75: Implement basic linkstate monitor task

### PR DESCRIPTION
**Pull request by lunkwill42**
This implements the basics of a link state monitor task, along with some of the necessary changes to the surrounding codebase.

As a side-effect, this also introduces basic test coverage for the zino.tasks module, which had none.

Closes #75

### CODE REVIEW

File: src/zino/tasks/linkstatetask.py [hunk 1]
```diff
@@ -0,0 +1,141 @@
...
+        state = f"admin{data.admin_status.capitalize()}"
+        # A special tweak so that we report ports in oper-down (but admin-up) state first time we see them
+        if not port.state and data.oper_status != "up" and state != "adminDown":
+            port.state = InterfaceState.UNKNOWN
+        if state == "adminUp":
+            state = data.oper_status
+        state = InterfaceState(state)
+        if port.state and port.state != state:
+            self._make_or_update_state_event(port, state)
+        port.state = state
```
Comment (stveit) [inline src/zino/tasks/linkstatetask.py:82]: I feel this is a bit hard to read, with `port.state` and `state` being used everywhere and set to different values several times. Even though its a pretty short function, I had to look at this pretty hard to understand whats happening

**Review by stveit — COMMENTED** [review]
Could increse coverage further by adding some more tests as codecov points out.

Would be nice with a couple small tests specifically for `BaseInterfaceRow.is_sane`

File: src/zino/tasks/linkstatetask.py (hunk 1 above)
Comment (lunkwill42) [reply src/zino/tasks/linkstatetask.py:82]: How do you think I felt when I read the original Tcl code this was translated from? :-D

File: src/zino/tasks/linkstatetask.py [hunk 2]
```diff
@@ -0,0 +1,141 @@
...
+        self._update_interfaces(attrs)
+
+    def _update_interfaces(self, new_attrs: SparseWalkResponse):
+        for index, row in new_attrs.items():
+            self._update_single_interface(row)
+
+    def _update_single_interface(self, row: dict[str, Any]):
+        data = BaseInterfaceRow(*(row.get(attr) for attr in BASE_POLL_LIST))
+        if not data.is_sane():
+            return
```
[the remark this answers is not in the document]
Comment (lunkwill42) [reply src/zino/tasks/linkstatetask.py:57]: If we collected data for 100 interfaces and a single one of them has faulty data, an exception could derail the correct update of the 99 other.  I agree with the logging part, but would like to stick as close to Zino 1's "algorithm" for "skipping" to the next interface.

File: src/zino/tasks/linkstatetask.py [hunk 3]
```diff
@@ -0,0 +1,141 @@
...
+        change = data.alias != port.ifalias
+
+        if change:
+            if not new:
+                _logger.info(
+                    "%s: changing desc for %s from %r to %r", self.device.name, data.index, port.ifalias, data.alias
+                )
+            else:
+                _logger.info("%s: setting desc for %s to %s", self.device.name, data.index, data.alias)
+            port.ifalias = data.alias
```
[the remark this answers is not in the document]
Comment (lunkwill42) [reply src/zino/tasks/linkstatetask.py:136]: Maybe, but being an old-sk00l coder I was taught to write the (presumed) most common path first in multi-branch statements: The list of interfaces won't change all that much after the first run, so I would presume the most common path is that existing interfaces are updated.

File: src/zino/tasks/linkstatetask.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (stveit) [reply src/zino/tasks/linkstatetask.py]: the exception can be handled in `_update_interfaces`, where you log it and go to the nest one. This way you get the same behaviour but more conventional code

**Review by stveit — APPROVED** [review]

**Pull request merged**